tools/applyconfig/applyconfig.go

```diff
@@ -0,0 +1,185 @@
+	flag.StringVar(&opt.level, "level", "standard", "Select which config to apply (standard, admin, all)")
+	flag.StringVar(&opt.user, "as", "", "Username to impersonate while applying the config")
+	flag.StringVar(&opt.directory, "config-dir", "", "Directory with config to apply")
+
+	flag.Parse()
+
+	if opt.level != "standard" && opt.level != "admin" && opt.level != "all" {
+		fmt.Fprintf(os.Stderr, "--level: must be one of [standard, admin, all]\n")
+		os.Exit(1)
+	}
+
+	if opt.directory == "" {
+		fmt.Fprintf(os.Stderr, "--config-dir must be provided\n")
+		os.Exit(1)
+	}
+
+	return opt
+}
+
+func isAdminConfig(filename string) bool {
+	return adminConfig.MatchString(filename)
+}
+
+func isStandardConfig(filename string) bool {
+	return filepath.Ext(filename) == ".yaml" &&
+		!isAdminConfig(filename)
+}
+
+func makeOcArgs(path, user string, dry bool) []string {
+	args := []string{"apply", "-f", path}
+	if dry {
+		args = append(args, "--dry-run")
+	}
+
+	if user != "" {
+		args = append(args, "--as", user)
+	}
+
+	return args
+}
+
+func apply(path, user string, dry bool) error {
+	args := makeOcArgs(path, user, dry)
+
+	cmd := exec.Command("oc", args...)
+	if output, err := cmd.CombinedOutput(); err != nil {
+		if _, ok := err.(*exec.ExitError); ok {
+			fmt.Printf("[ERROR] oc %s: failed to apply\n%s\n", strings.Join(args, " "), string(output))
+		} else {
+			fmt.Printf("[ERROR] oc %s: failed to execute: %v\n", strings.Join(args, " "), err)
+		}
+		return fmt.Errorf("failed to apply config")
+	}
+
+	fmt.Printf("oc %s: OK\n", strings.Join(args, " "))
+	return nil
+}
+
+func applyAdminConfig(rootDir, user string, dry bool) error {
+	failures := false
+	if user == "" {
+		user = "system:admin"
+	}
+	if err := filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			if strings.HasPrefix(info.Name(), "_") {
+				fmt.Printf("Skipping directory: %s\n", path)
+				return filepath.SkipDir
+			}
+			fmt.Printf("Applying admin config in directory: %s\n", path)
+			return nil
+		}
+
+		if isAdminConfig(info.Name()) {
+			if err := apply(path, user, dry); err != nil {
+				failures = true
+			}
+		}
+
+		return nil
+	}); err != nil {
+		// should not happen
+		fmt.Fprintf(os.Stderr, "failed to walk directory '%s': %v\n", rootDir, err)
+		return err
+	}
+
+	if failures {
+		return fmt.Errorf("failed to apply admin config")
+	}
+
+	return nil
+}
+
+func applyStandardConfig(rootDir, user string, dry bool) error {
+	failures := false
+	if err := filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			if strings.HasPrefix(info.Name(), "_") {
+				fmt.Printf("Skipping directory: %s\n", path)
+				return filepath.SkipDir
+			}
+			fmt.Printf("Applying config in directory: %s\n", path)
+			return nil
+		}
+
+		if isStandardConfig(info.Name()) {
```

Comment:
Could you DRY out the `walk` to just take the function to call here as a parameter?

Reply:
Yeah. I tried to DRY it in one of the earlier iterations, it was hard so gave up. In the current form it should not be hard anymore.